compact_bytes/src/fixed.rs

```diff
@@ -0,0 +1,471 @@
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use compact_bytes::CompactBytesSlice;
+    ///
+    /// let min = CompactBytesSlice::empty();
+    /// assert_eq!(min.capacity(), CompactBytesSlice::MAX_INLINE);
+    /// ```
+    #[inline]
+    pub const fn empty() -> Self {
+        let inline = InlineBytes15::empty();
+        CompactBytesSlice { inline }
+    }
+
+    /// Creates a new [`CompactBytesSlice`] using the provided pointer and length.
+    ///
+    /// # Safety
+    ///
+    /// * The caller must guarantee that the provided pointer is properly aligned, and the backing
+    ///   allocation was made by the same allocator that will eventually be used to free the
+    ///   returned [`CompactBytesSlice`].
+    /// * `length` needs to be the capacity that the pointer was allocated with.
+    /// * `length` needs to be less than or equal to [`CompactBytesSlice::MAX_HEAP`].
+    ///
+    #[inline]
+    pub unsafe fn from_raw_parts(ptr: *mut u8, length: usize) -> Self {
+        let heap = HeapBytesFixed {
+            ptr: NonNull::new_unchecked(ptr),
+            len: length,
+        };
+        let heap = ManuallyDrop::new(heap);
+        CompactBytesSlice { heap }
+    }
+
+    /// Returns the contents of the [`CompactBytesSlice`] as a bytes slice.
+    #[inline]
+    pub fn as_slice(&self) -> &[u8] {
+        let pointer = self.as_ptr();
+        let length = self.len();
+
+        unsafe { core::slice::from_raw_parts(pointer, length) }
+    }
+
+    /// Returns the contents of the [`CompactBytesSlice`] as a mutable bytes slice.
+    #[inline]
+    pub fn as_mut_slice(&mut self) -> &mut [u8] {
+        let pointer = self.as_mut_ptr();
+        let length = self.len();
+
+        unsafe { core::slice::from_raw_parts_mut(pointer, length) }
+    }
+
+    /// Returns the length of the [`CompactBytesSlice`].
+    #[inline(always)]
+    pub fn len(&self) -> usize {
+        // SAFETY: `InlineBytes15` and `HeapBytesFixed` have the same size and alignment.
+        //
+        // Note: This code is very carefully written so we can benefit from branchless
+        // instructions.
+        let (mut length, heap_length) = unsafe { (self.inline.len(), self.heap.len) };
+        if self.spilled() {
+            length = heap_length;
+        }
+
+        length
+    }
+
+    /// Returns if the [`CompactBytesSlice`] is empty.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns the capacity of the [`CompactBytesSlice`].
+    #[inline(always)]
+    pub fn capacity(&self) -> usize {
+        // SAFETY: `InlineBytes15` and `HeapBytesFixed` share the same size and alignment.
+        //
+        // Note: This code is very carefully written so we can benefit from branchless
+        // instructions.
+        let (mut capacity, heap_capacity) = unsafe { (Self::MAX_INLINE, self.heap.len) };
+        if self.spilled() {
+            capacity = heap_capacity;
+        }
+        capacity
+    }
```

Comment:
Should we have a `capacity`? It's always equal to `len`. If we decide to keep it, I think it should just call `len`.

Reply:
I re-added `CompactBytesSlice::capacity` because when it's an inline variant the capacity != len, so it might be useful to keep around